include/caffe/layer.hpp

```diff
@@ -413,7 +413,7 @@ template <typename Dtype>
 inline Dtype Layer<Dtype>::Forward(const vector<Blob<Dtype>*>& bottom,
     const vector<Blob<Dtype>*>& top) {
   Dtype loss = 0;
-  Reshape(bottom, top);
+  // Reshape(bottom, top);
```

Comment:
The different size inputs of a batch should call net->reshape() explicitly to update all networking layers, instead of doing this magically by Forward(). There are two shortages to have a call reshape just before forward(): 1. bring overhead to forward for cases that have no shape change. 2. If reshape have allocated some resource, it will have to handle two different situations for the second call in Forward(): the shape changes and it is necessary to release and reallocate resource, or the shape does not change and it should not do resource reclaim and allocation.

It is really confusing to have a Reshape() in Forward, based on Reshape()'s semantic.

Thanks,

Cyber